File: pillpal_app/lib/models/medicine.dart
import 'package:cloud_firestore/cloud_firestore.dart';

class Medicine {
  final String id;
  final String userId;
  final String memberName;
  final String name;
  final String dosage;
  final String scheduledTime;
  final String frequency;
  final List<int> daysOfWeek;
  final int pillCount;
  final int refillAt;
  final bool active;
  final DateTime createdAt;

  Medicine({
    required this.id,
    required this.userId,
    this.memberName = 'Self',
    required this.name,
    required this.dosage,
    required this.scheduledTime,
    this.frequency = 'daily',
    this.daysOfWeek = const [],
    this.pillCount = 30,
    this.refillAt = 5,
    this.active = true,
    required this.createdAt,
  });

  factory Medicine.fromDoc(DocumentSnapshot doc) {
    final data = doc.data() as Map<String, dynamic>;
    return Medicine(
      id: doc.id,
      userId: data['userId'] ?? '',
      memberName: data['memberName'] ?? 'Self',
      name: data['name'] ?? '',
      dosage: data['dosage'] ?? '',
      scheduledTime: data['scheduledTime'] ?? '08:00',
      frequency: data['frequency'] ?? 'daily',
      daysOfWeek: List<int>.from(data['daysOfWeek'] ?? []),
      pillCount: data['pillCount'] ?? 30,
      refillAt: data['refillAt'] ?? 5,
      active: data['active'] ?? true,
      createdAt: (data['createdAt'] as Timestamp?)?.toDate() ?? DateTime.now(),
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'userId': userId,
      'memberName': memberName,
      'name': name,
      'dosage': dosage,
      'scheduledTime': scheduledTime,
      'frequency': frequency,
      'daysOfWeek': daysOfWeek,
      'pillCount': pillCount,
      'refillAt': refillAt,
      'active': active,
      'createdAt': FieldValue.serverTimestamp(),
    };
  }

  Medicine copyWith({int? pillCount, bool? active}) {
    return Medicine(
      id: id,
      userId: userId,
      memberName: memberName,
      name: name,
      dosage: dosage,
      scheduledTime: scheduledTime,
      frequency: frequency,
      daysOfWeek: daysOfWeek,
      pillCount: pillCount ?? this.pillCount,
      refillAt: refillAt,
      active: active ?? this.active,
      createdAt: createdAt,
    );
  }
}

File: backend/app/models/medicine.py
from __future__ import annotations

import uuid
from datetime import datetime, time, timezone

from sqlalchemy import Boolean, DateTime, ForeignKey, Integer, String, Time
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class Medicine(Base):
    __tablename__ = "medicines"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), index=True
    )
    name: Mapped[str] = mapped_column(String(255))
    dosage: Mapped[str] = mapped_column(String(255))
    scheduled_time: Mapped[time] = mapped_column(Time)
    frequency: Mapped[str] = mapped_column(String(32), default="daily")
    active: Mapped[bool] = mapped_column(Boolean, default=True)
    pill_count: Mapped[int | None] = mapped_column(Integer, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )

    user: Mapped["User"] = relationship("User", back_populates="medicines")
    dose_logs: Mapped[list["DoseLog"]] = relationship(
        "DoseLog", back_populates="medicine", cascade="all, delete-orphan"
    )
